✨ Licence compliance check

acabarbaye opened the pull request:
### Description
This PR is for checking that the licences of a package and its dependencies are compliant with a list of authorised licences which is defined in the project configuration. When reporting the licences (generating the SPDX documents), the compliance of all licences is done and if one licence is not in the list set, the action will fail.
A big problem of licences is that their descriptor in Python packages is free form text. Therefore, an Apache-2 package can have the different licence descriptors below to say so:
           - "Apache 2.0",
           - "Apache Licence, Version 2.0",
           - "Apache 2",
           - "Apache License Version 2.0",
           - "Apache License, Version 2.0",
           - "Apache License 2.0",
            - "Apache-2",
            - "Apache-2.0"
Therefore, I had to implement some string comparison algorithm in order to find out which actual licence a descriptor is referring to so that I can see if such licence is compliant or not.

#### Note:
I apologise in advance for the size of this PR. A lot of it is a result of some refactoring. More to come in order to generate some third party IP summaries


For information, this is the list of cases, I had to deal with for this package:
**__python package__**: _licence reported_ 
```
dataclasses: Apache
jellyfish: UNKNOWN
wcmatch: MIT License
license-expression: apache-2.0
spdx-tools: Apache-2.0
licenseheaders: MIT
packaging: BSD or Apache License, Version 2.0
boto3: Apache License 2.0
twine: UNKNOWN
python-dotenv: UNKNOWN
toml: MIT
pdoc3: AGPL-3.0
pyautoversion: Apache 2.0
towncrier: MIT
GitPython: UNKNOWN
backrefs: MIT License
bracex: MIT License
boolean.py: BSD-2-Clause
PyYAML: MIT
rdflib: BSD-3-Clause
xmltodict: MIT
ply: BSD
six: MIT
regex: Python Software Foundation License
pyparsing: MIT License
s3transfer: Apache License 2.0
botocore: Apache License 2.0
jmespath: MIT
requests: Apache 2.0
importlib-metadata: Apache Software License
readme-renderer: Apache License, Version 2.0
setuptools: UNKNOWN
tqdm: MPLv2.0, MIT Licences
pkginfo: MIT
requests-toolbelt: Apache 2.0
keyring: UNKNOWN
Mako: MIT
Markdown: BSD License
semver: BSD
incremental: MIT
Jinja2: BSD-3-Clause
click: BSD-3-Clause
gitdb: BSD License
isodate: BSD
pyparsing: MIT License
botocore: Apache License 2.0
urllib3: MIT
python-dateutil: Dual License
certifi: MPL-2.0
idna: BSD-like
chardet: LGPL
urllib3: MIT
zipp: UNKNOWN
Pygments: BSD License
bleach: Apache Software License
docutils: public domain, Python, 2-Clause BSD, GPL 3 (see COPYING.txt)
requests: Apache 2.0
pywin32-ctypes: BSD
MarkupSafe: BSD-3-Clause
setuptools: UNKNOWN
MarkupSafe: BSD-3-Clause
smmap: BSD
six: MIT
webencodings: BSD
mbed-tools-ci-scripts: Apache 2.0
```


### Test Coverage
- [x]  This change is covered by existing or additional automated tests.
- [x]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

acabarbaye committed:
creating a way to identify equivalent strings

acabarbaye committed:
Added unit tests

acabarbaye committed:
modified spdx report to use licence lookup

acabarbaye committed:
modified configuration to list accepted licences

acabarbaye committed:
checking licence compliance of a package and its dependencies

acabarbaye committed:
copyright and news file

acabarbaye committed:
making code climate happy

acabarbaye committed:
making code climate even more happy and hopefully for good

acabarbaye committed:
formatting improvement

acabarbaye committed:
try forcing publishing artifacts even if check fails

rwalton-arm commented on mbed_tools_ci_scripts/report_third_party_ip.py:
```diff
@@ -51,7 +53,8 @@ def main() -> int:
 
     try:
-        generate_spdx_reports(Path(args.output_dir))
+        project = generate_spdx_reports(Path(args.output_dir))
```
If we pass a callable that converts `args.output_dir` to a `pathlib.Path` to the `type` kwarg of `parser.add_argument` we wouldn't have to do the conversion here. That might be a bit cleaner? We only use this once AFAICT so it's not that important at the moment.

acabarbaye replied:
ok. will do

acabarbaye replied:
418326abd2b04ddfb63cea5c08fe58e7951442d4

rwalton-arm commented on mbed_tools_ci_scripts/spdx_report/spdx_document.py:
```diff
@@ -177,6 +177,11 @@ def external_refs(self) -> List[DependencySpdxDocumentRef]:
         """
         return self._other_document_references
 
+    @external_refs.setter
```
Do we really need this to be a property? There's no encapsulation here, as we just pass `external_refs` directly to `_other_document_references`. Would it be better to just have `external_refs` as a public attribute and remove `_other_document_references`?

acabarbaye replied:
True. I thought nevertheless that it made the documentation clearer with proper docstring for the field.
![image](https://user-images.githubusercontent.com/29707928/79448551-85f0c100-7fd9-11ea-8c33-4d153409242a.png)

rwalton-arm commented on mbed_tools_ci_scripts/spdx_report/spdx_file.py:
```diff
@@ -116,8 +117,8 @@ def generate_spdx_file(self) -> File:
         source_file.comment = determine_spdx_value(None)
         source_file.chk_sum = Algorithm("SHA1", self.sha1_check_sum)
-        source_file.conc_lics = License.from_identifier(str(determine_spdx_value(self.licence)))
+        source_file.conc_lics = License.from_identifier(str(determine_spdx_value(self.licence.identifier)))
```
I think this attribute could do with a better name. I'm not sure what a `conc_lics` is.

acabarbaye replied:
this is a field defined by the SPDX SDK. I don't have control over the names

rwalton-arm commented on mbed_tools_ci_scripts/spdx_report/spdx_file.py:
```diff
@@ -116,8 +117,8 @@ def generate_spdx_file(self) -> File:
         source_file.spdx_id = f"SPDXRef-{self.id}"
         source_file.copyright = determine_spdx_value(self.copyright)
-        source_file.add_lics(License.from_identifier(str(determine_spdx_value(self.licence))))
+        source_file.add_lics(License.from_identifier(str(determine_spdx_value(self.licence.identifier))))
```
I'm not sure why we have two ways of setting the `License` on the `source_file`? We monkey patch `conc_lics` above, then we also have this `add_lics` method. 😕

acabarbaye replied:
this is the weirdness of the SDK I have at disposal I am afraid.

rwalton-arm commented on mbed_tools_ci_scripts/spdx_report/spdx_package.py:
```diff
@@ -206,7 +223,7 @@ def generate_spdx_package(self) -> Package:
         package.homepage = determine_spdx_value(self.url)
         package.originator = Person(determine_spdx_value(self.author), determine_spdx_value(self.author_email))
-        package.license_declared = License.from_identifier(str(determine_spdx_value(self.main_licence)))
+        package.license_declared = License.from_identifier(str(determine_spdx_value(self.main_licence.identifier)))
```
Would it make more sense to set these attributes in `Package.__init__` rather than monkey patching? We could just pass a reference to `self` to the `Package` and then set the attributes when we construct `Package`. It would make this calling code easier to follow.

acabarbaye replied:
I did what I could. Package is a class provided by the SDK and the __init__ does not have all the attributes defined. Only set the one I could

acabarbaye replied:
418326abd2b04ddfb63cea5c08fe58e7951442d4

rwalton-arm commented on mbed_tools_ci_scripts/utils/string_helpers.py:
```diff
@@ -0,0 +1,68 @@
+def determine_similar_string_from_list(string: str, strings: Iterable[str]) -> str:
+    """Determines the closest string to a string from a list."""
+    string_to_assess = string.strip()
+    return (sorted([determine_matching_stats(string_to_assess, s) for s in strings])[0]).string2
```
Could we unpack this expression a bit? It's kind of hard to follow in it's current form.

Something like:
```
sorted_strings = sorted([determine_matching_stats(string_to_assess, s) for s in strings])
return sorted_strings[0].string2
```

acabarbaye replied:
bdd6785547f28ab2559a7a11346fb4007fb65a86

rwalton-arm commented on mbed_tools_ci_scripts/utils/third_party_licences.py:
```diff
@@ -0,0 +1,139 @@
+    )
+
+
+def parse_licences(licence_info: dict) -> Iterable[Licence]:
```
It might be clearer to name this something like `iter_licenses` so it's obvious that it's a generator function.

acabarbaye replied:
will do

acabarbaye replied:
418326abd2b04ddfb63cea5c08fe58e7951442d4

rwalton-arm commented on mbed_tools_ci_scripts/utils/third_party_licences.py:
```diff
@@ -0,0 +1,139 @@
+def is_licence_accepted(licence_expression: str) -> bool:
+    """Determines whether the licence expressed is valid with regards to project's accepted licences."""
+    authorised_licences = [l.identifier for l in get_allowed_opensource_licences()]
+    return len([l for l in determine_licences_not_in_list(licence_expression, iter(authorised_licences))]) == 0
```
```suggestion
    return bool([l for l in determine_licences_not_in_list(licence_expression, iter(authorised_licences))])
```

acabarbaye replied:
418326abd2b04ddfb63cea5c08fe58e7951442d4

acabarbaye committed:
Update mbed_tools_ci_scripts/utils/third_party_licences.py

Co-Authored-By: Rob Walton <rob.walton@arm.com>

acabarbaye committed:
changes following review comments

acabarbaye committed:
Updated the PR to handle more licence descriptor cases.

acabarbaye committed:
Adding a way to use a regex to defined allowed licences e.g. `BSD* `

acabarbaye committed:
licence expression understanding improvement

rwalton-arm commented on mbed_tools_ci_scripts/utils/third_party_licences.py:
```diff
@@ -90,8 +92,19 @@ def get_licence(self, licence_descriptor: Optional[str]) -> Optional[Licence]:
-        # TODO refine the licence_descriptor: remove The at begining of strings.
-        return self._licence_store.get(determine_similar_string_from_list(licence_descriptor, self._licence_list))
+        cleansed_descriptor = cleanse_licence_descriptor(licence_descriptor)
+        likelihood, licence = determine_similar_string_from_list(cleansed_descriptor, self._licence_list)
```
I'm wondering if it might be better to push the "likelihood" check down into `determine_similar_string_from_list`? That way we don't have to return a `tuple` and the function has a clearer post-condition, i.e it returns `None` if there's no similar string in the list, or it just returns the similar string. At the moment it's returning `candidate_similar_string_from_list` and passing the responsibility of determining whether that is a similar string or not to the caller. Thoughts?

acabarbaye replied:
The reason I did it like that is that the threshold value is really dependent on the kind of strings we are dealing with. For licences, a threshold of 0.5 seems to be good enough. However, for other types of strings, this may be completely erroneous. As these helpers are meant to be generic enough so that they can be reused elsewhere, I did not want to make too bold decisions within the helper method but rather let the responsibility to make them where we have more context about the kind of strings we are dealing with

acabarbaye committed:
fixing problems with comma in the licence description

acabarbaye committed:
fix regex

acabarbaye committed:
fix test

acabarbaye committed:
Better handled licences and put in place a way to mark packages as checked with regards to licensing

acabarbaye committed:
listing the packages for which the licence has been checked
